File: backend/routes/indoor_plants.py
from flask import Blueprint, request, jsonify
from models.indoor_plant import IndoorPlant
from app import db

indoor_plants_bp = Blueprint('indoor_plants', __name__, url_prefix='/indoor-plants')

@indoor_plants_bp.route('/', methods=['POST'])
def create_indoor_plant():
    data = request.get_json()
    
    if not data:
        return jsonify({'error': 'No data provided'}), 400
    
    if not data.get('scientific_name'):
        return jsonify({'error': 'Scientific name is required'}), 400
    
    existing_plant = IndoorPlant.query.filter_by(
        scientific_name=data.get('scientific_name')
    ).first()
    if existing_plant:
        return jsonify({'error': 'Plant with this scientific name already exists'}), 409
    
    plant = IndoorPlant(
        scientific_name=data.get('scientific_name'),
        common_names=data.get('common_names'),
        family=data.get('family'),
        origin=data.get('origin'),
        difficulty=data.get('difficulty'),
        watering_frequency=data.get('watering_frequency'),
        light=data.get('light'),
        humidity=data.get('humidity'),
        temperature=data.get('temperature'),
        soil_type=data.get('soil_type'),
        adult_size=data.get('adult_size'),
        growth_rate=data.get('growth_rate'),
        toxicity=data.get('toxicity'),
        air_purification=data.get('air_purification', False),
        flowering=data.get('flowering')
    )
    
    try:
        db.session.add(plant)
        db.session.commit()
        return jsonify(plant.to_dict()), 201
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': 'Failed to create plant'}), 500

@indoor_plants_bp.route('/', methods=['GET'])
def list_indoor_plants():
    query = IndoorPlant.query
    search = request.args.get('search')
    if search:
        query = query.filter(
            IndoorPlant.scientific_name.ilike(f'%{search}%') |
            IndoorPlant.common_names.ilike(f'%{search}%')
        )
    
    difficulty = request.args.get('difficulty')
    if difficulty:
        query = query.filter(IndoorPlant.difficulty == difficulty)
    
    family = request.args.get('family')
    if family:
        query = query.filter(IndoorPlant.family == family)
    
    plants = query.all()
    return jsonify([p.to_dict() for p in plants]), 200

@indoor_plants_bp.route('/<int:plant_id>', methods=['GET'])
def get_indoor_plant(plant_id):
    plant = IndoorPlant.query.get_or_404(plant_id)
    return jsonify(plant.to_dict()), 200

@indoor_plants_bp.route('/<int:plant_id>', methods=['PUT'])
def update_indoor_plant(plant_id):
    plant = IndoorPlant.query.get_or_404(plant_id)
    data = request.get_json()
    
    if not data:
        return jsonify({'error': 'No data provided'}), 400
    
    for key, value in data.items():
        if hasattr(plant, key):
            setattr(plant, key, value)
    
    db.session.commit()
    return jsonify(plant.to_dict()), 200

@indoor_plants_bp.route('/<int:plant_id>', methods=['DELETE'])
def delete_indoor_plant(plant_id):
    plant = IndoorPlant.query.get_or_404(plant_id)
    db.session.delete(plant)
    db.session.commit()
    return jsonify({'message': 'Plant deleted successfully'}), 200

File: frontend/src/stores/indoorPlants.ts
import { defineStore } from 'pinia'
import { fetchIndoorPlants, createIndoorPlant, updateIndoorPlant, deleteIndoorPlant } from '../services/api'
import type { IndoorPlant } from '../types'


export const useIndoorPlantsStore = defineStore('indoorPlants', {
  state: () => ({
    plants: [] as IndoorPlant[],
    loading: false,
    error: null as string | null
  }),
  actions: {
    async fetchPlants(search?: string) {
      this.loading = true
      try {
        const backendPlants = await fetchIndoorPlants(search)
        this.plants = backendPlants.map((p: any) => ({
          id: p.id,
          scientific_name: p.scientific_name,
          common_names: p.common_names,
          family: p.family,
          difficulty: p.difficulty,
          origin: p.origin,
          watering_frequency: p.watering_frequency,
          light: p.light,
          humidity: p.humidity,
          temperature: p.temperature,
          soil_type: p.soil_type,
          adult_size: p.adult_size,
          growth_rate: p.growth_rate,
          toxicity: p.toxicity,
          air_purification: p.air_purification,
          flowering: p.flowering
        }))
        this.error = null
      } catch (e: any) {
        this.error = e.message
      } finally {
        this.loading = false
      }
    },
    async addPlant(payload: Omit<IndoorPlant, 'id' | 'created_at' | 'updated_at'>) {
      this.loading = true
      try {
        // Appel API direct avec le payload aligné
        const backendPlant = await createIndoorPlant(payload)
        const plant: IndoorPlant = {
          id: backendPlant.id,
          scientific_name: backendPlant.scientific_name,
          common_names: backendPlant.common_names,
          family: backendPlant.family,
          difficulty: backendPlant.difficulty,
          origin: backendPlant.origin,
          watering_frequency: backendPlant.watering_frequency,
          light: backendPlant.light,
          humidity: backendPlant.humidity,
          temperature: backendPlant.temperature,
          soil_type: backendPlant.soil_type,
          adult_size: backendPlant.adult_size,
          growth_rate: backendPlant.growth_rate,
          toxicity: backendPlant.toxicity,
          air_purification: backendPlant.air_purification,
          flowering: backendPlant.flowering
        }
        this.plants.unshift(plant)
        this.error = null
      } catch (e: any) {
        this.error = e.message
      } finally {
        this.loading = false
      }
    },
    async updatePlant(id: number, payload: Partial<Omit<IndoorPlant, 'id' | 'created_at' | 'updated_at'>>) {
      this.loading = true
      try {
        const updated = await updateIndoorPlant(id, payload) as IndoorPlant
        const idx = this.plants.findIndex((p) => p.id === id)
        if (idx !== -1) this.plants[idx] = updated
        this.error = null
      } catch (e: any) {
        this.error = e.message
      } finally {
        this.loading = false
      }
    },
    async deletePlant(id: number) {
      this.loading = true
      try {
        await deleteIndoorPlant(id)
        this.plants = this.plants.filter((p) => p.id !== id)
        this.error = null
      } catch (e: any) {
        this.error = e.message
      } finally {
        this.loading = false
      }
    }
  }
})
